sklearn/decomposition/nmf.py

```diff
@@ -254,22 +344,287 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+             - safe_sparse_dot(X, H.T, dense_output=True))
+    gradH = (np.dot(np.dot(W.T, W), H)
+             - safe_sparse_dot(W.T, X, dense_output=True))
+
+    init_grad = squared_norm(gradW) + squared_norm(gradH.T)
+    # max(0.001, tol) to force alternating minimizations of W and H
+    tolW = max(0.001, tol) * np.sqrt(init_grad)
+    tolH = tolW
+
+    for n_iter in range(1, max_iter + 1):
+        # stopping condition
+        # as discussed in paper
+        proj_grad_W = squared_norm(gradW * np.logical_or(gradW < 0, W > 0))
+        proj_grad_H = squared_norm(gradH * np.logical_or(gradH < 0, H > 0))
+
+        if (proj_grad_W + proj_grad_H) / init_grad < tol ** 2:
+            break
+
+        # update W
+        W, gradW, iterW = _update_projected_gradient_w(X, W, H, tolW,
+                                                       nls_max_iter,
+                                                       alpha, l1_ratio,
+                                                       sparseness, beta, eta)
+        if iterW == 1:
+            tolW = 0.1 * tolW
+
+        # update H
+        H, gradH, iterH = _update_projected_gradient_h(X, W, H, tolH,
+                                                       nls_max_iter,
+                                                       alpha, l1_ratio,
+                                                       sparseness, beta, eta)
+        if iterH == 1:
+            tolH = 0.1 * tolH
+
+    H[H == 0] = 0   # fix up negative zeros
+
+    if n_iter == max_iter:
+        W, _, _ = _update_projected_gradient_w(X, W, H, tol, nls_max_iter,
+                                               alpha, l1_ratio, sparseness,
+                                               beta, eta)
+
+    return W, H, n_iter
+
+
+def _update_coordinate_descent(X, W, Ht, alpha, l1_ratio, shuffle,
+                               random_state):
+    """Helper function for _fit_coordinate_descent"""
+    n_components = Ht.shape[1]
+
+    HHt = fast_dot(Ht.T, Ht)
+    XHt = safe_sparse_dot(X, Ht)
+
+    # L1 and L2 regularizations
+    l1_reg = 1. * l1_ratio * alpha
+    l2_reg = (1. - l1_ratio) * alpha
+
+    # L2 regularization corresponds to increase the diagonal of HHt
+    if l2_reg != 0.:
+        HHt += np.eye(n_components) * l2_reg
+    # L1 regularization correponds to decrease each element of XHt
+    if l1_reg != 0.:
+        XHt -= l1_reg
+
+    seed = random_state.randint(np.iinfo(np.int32).max)
+    return _update_cdnmf_fast(W, HHt, XHt, shuffle, seed)
+
+
+def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+                            l1_ratio=0., regularization=None, update_H=True,
+                            verbose=0, shuffle=False, random_state=None):
+    """Compute Non-negative Matrix Factorization (NMF) with Coordinate Descent
+
+    Find (W, H) that minimize || X - W.H ||_2
+    with an alternative least square minimization of W and H.
+    Each minimization is done with a Cyclic Coordinate Descent.
```

Comment:
Also I'm not sure whether `alternative least square minimization` is clearer than `alternative minimization` here.